File: splp-go/pkg/crypto/encryption.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"

	"github.com/perlinsos/splp-go/pkg/types"
)

const (
	// AES-256-GCM constants
	algorithm = "aes-256-gcm"
	keySize   = 32 // 256 bits
	ivLength  = 16 // 128 bits (16 bytes) - matching TypeScript
	tagLength = 16 // 128 bits
)

// Encryptor implements the encryption interface using AES-256-GCM
type Encryptor struct {
	key []byte
}

// NewEncryptor creates a new encryptor with the provided hex key
func NewEncryptor(hexKey string) (*Encryptor, error) {
	key, err := hex.DecodeString(hexKey)
	if err != nil {
		return nil, fmt.Errorf("invalid hex key: %w", err)
	}

	if len(key) != keySize {
		return nil, fmt.Errorf("encryption key must be %d bytes (64 hex characters) for AES-256", keySize)
	}

	return &Encryptor{key: key}, nil
}

// EncryptPayload encrypts a JSON payload using AES-256-GCM
// This matches the TypeScript encryptPayload function exactly
func (e *Encryptor) EncryptPayload(payload interface{}, requestID string) (*types.EncryptedMessage, error) {
	// Convert payload to JSON string
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	// Create AES cipher
	block, err := aes.NewCipher(e.key)
	if err != nil {
		return nil, fmt.Errorf("failed to create AES cipher: %w", err)
	}

	// Create GCM mode with custom nonce size for 16-byte IV
	gcm, err := cipher.NewGCMWithNonceSize(block, ivLength)
	if err != nil {
		return nil, fmt.Errorf("failed to create GCM: %w", err)
	}

	// Generate random IV (16 bytes to match TypeScript)
	iv := make([]byte, ivLength)
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return nil, fmt.Errorf("failed to generate IV: %w", err)
	}

	// Encrypt the data - in Go, gcm.Seal appends the tag to the ciphertext
	// But TypeScript keeps them separate, so we need to split them
	sealedData := gcm.Seal(nil, iv, payloadBytes, nil)

	// The last 16 bytes are the authentication tag
	if len(sealedData) < tagLength {
		return nil, fmt.Errorf("sealed data too short")
	}

	encryptedData := sealedData[:len(sealedData)-tagLength]
	tag := sealedData[len(sealedData)-tagLength:]

	return &types.EncryptedMessage{
		RequestID: requestID, // Not encrypted for tracing
		Data:      hex.EncodeToString(encryptedData),
		IV:        hex.EncodeToString(iv),
		Tag:       hex.EncodeToString(tag),
	}, nil
}

// DecryptPayload decrypts an encrypted message using AES-256-GCM
// This matches the TypeScript decryptPayload function exactly
func (e *Encryptor) DecryptPayload(encryptedMessage *types.EncryptedMessage) (string, interface{}, error) {
	// Convert IV and tag from hex
	iv, err := hex.DecodeString(encryptedMessage.IV)
	if err != nil {
		return "", nil, fmt.Errorf("failed to decode IV: %w", err)
	}

	tag, err := hex.DecodeString(encryptedMessage.Tag)
	if err != nil {
		return "", nil, fmt.Errorf("failed to decode tag: %w", err)
	}

	// Convert encrypted data from hex
	encryptedData, err := hex.DecodeString(encryptedMessage.Data)
	if err != nil {
		return "", nil, fmt.Errorf("failed to decode encrypted data: %w", err)
	}

	// Create AES cipher
	block, err := aes.NewCipher(e.key)
	if err != nil {
		return "", nil, fmt.Errorf("failed to create AES cipher: %w", err)
	}

	// Create GCM mode with custom nonce size for 16-byte IV
	gcm, err := cipher.NewGCMWithNonceSize(block, ivLength)
	if err != nil {
		return "", nil, fmt.Errorf("failed to create GCM: %w", err)
	}

	// Combine encrypted data and tag for decryption (gcm.Open expects this format)
	sealedData := append(encryptedData, tag...)

	// Decrypt the data
	plaintext, err := gcm.Open(nil, iv, sealedData, nil)
	if err != nil {
		return "", nil, fmt.Errorf("failed to decrypt: %w", err)
	}

	// Parse JSON payload
	var payload interface{}
	if err := json.Unmarshal(plaintext, &payload); err != nil {
		return "", nil, fmt.Errorf("failed to unmarshal decrypted payload: %w", err)
	}

	return encryptedMessage.RequestID, payload, nil
}

// Legacy methods for backward compatibility
// Encrypt is an alias for EncryptPayload
func (e *Encryptor) Encrypt(payload interface{}, requestID string) (*types.EncryptedMessage, error) {
	return e.EncryptPayload(payload, requestID)
}

// Decrypt is an alias for DecryptPayload
func (e *Encryptor) Decrypt(encryptedMessage *types.EncryptedMessage) (string, interface{}, error) {
	return e.DecryptPayload(encryptedMessage)
}

// GenerateEncryptionKey generates a random encryption key for AES-256
// This matches the TypeScript generateEncryptionKey function exactly
func GenerateEncryptionKey() string {
	key := make([]byte, keySize)
	if _, err := io.ReadFull(rand.Reader, key); err != nil {
		panic(fmt.Sprintf("failed to generate encryption key: %v", err))
	}
	return hex.EncodeToString(key)
}

// ValidateKey validates that a hex key is the correct length for AES-256
func ValidateKey(hexKey string) error {
	key, err := hex.DecodeString(hexKey)
	if err != nil {
		return fmt.Errorf("invalid hex key: %w", err)
	}

	if len(key) != keySize {
		return fmt.Errorf("encryption key must be %d bytes (64 hex characters) for AES-256", keySize)
	}

	return nil
}

File: splp-bun/src/lib/crypto/encryption.ts
import { createCipheriv, createDecipheriv, randomBytes } from 'crypto';
import type { EncryptedMessage } from '../../types/index.js';

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 12; // Changed from 16 to 12 for .NET AesGcm compatibility (96 bits)
const TAG_LENGTH = 16;

/**
 * Encrypts a JSON payload using AES-256-GCM
 * @param payload - The data to encrypt
 * @param encryptionKey - 32-byte hex string for AES-256
 * @param requestId - The request ID (not encrypted, used for correlation)
 * @returns Encrypted message with IV and auth tag
 */
export function encryptPayload<T>(
  payload: T,
  encryptionKey: string,
  requestId: string
): EncryptedMessage {
  // Convert hex key to buffer
  const key = Buffer.from(encryptionKey, 'hex');

  if (key.length !== 32) {
    throw new Error('Encryption key must be 32 bytes (64 hex characters) for AES-256');
  }

  // Generate random IV
  const iv = randomBytes(IV_LENGTH);

  // Convert payload to JSON string
  const payloadString = JSON.stringify(payload);

  // Create cipher
  const cipher = createCipheriv(ALGORITHM, key, iv);

  // Encrypt the data
  let encrypted = cipher.update(payloadString, 'utf8', 'hex');
  encrypted += cipher.final('hex');

  // Get authentication tag
  const tag = cipher.getAuthTag();

  return {
    request_id: requestId, // Not encrypted for tracing
    data: encrypted,
    iv: iv.toString('hex'),
    tag: tag.toString('hex'),
  };
}

/**
 * Decrypts an encrypted message using AES-256-GCM
 * @param encryptedMessage - The encrypted message
 * @param encryptionKey - 32-byte hex string for AES-256
 * @returns Decrypted payload
 */
export function decryptPayload<T>(
  encryptedMessage: EncryptedMessage,
  encryptionKey: string
): { requestId: string; payload: T } {
  // Convert hex key to buffer
  const key = Buffer.from(encryptionKey, 'hex');

  if (key.length !== 32) {
    throw new Error('Encryption key must be 32 bytes (64 hex characters) for AES-256');
  }

  // Convert IV and tag from hex
  const iv = Buffer.from(encryptedMessage.iv, 'hex');
  const tag = Buffer.from(encryptedMessage.tag, 'hex');

  // Create decipher
  const decipher = createDecipheriv(ALGORITHM, key, iv);
  decipher.setAuthTag(tag);

  // Decrypt the data
  let decrypted = decipher.update(encryptedMessage.data, 'hex', 'utf8');
  decrypted += decipher.final('utf8');

  // Parse JSON payload
  const payload = JSON.parse(decrypted) as T;

  return {
    requestId: encryptedMessage.request_id,
    payload,
  };
}

/**
 * Generates a random encryption key for AES-256
 * @returns 32-byte hex string
 */
export function generateEncryptionKey(): string {
  return randomBytes(32).toString('hex');
}
